### PR TITLE
adding tfa2017 protocol with unit test

### PR DESCRIPTION
**Pull request by behrisch**
The TFA protocol with a unit test against the rewrite branch.

### CODE REVIEW

**Comment by CurlyMoo** [comment]
Can you sync with the latest code so the unittest actually succeeds?

**Comment by behrisch** [comment]
The test which fails is not the one you fixed, so I think it won't help.

File: libs/pilight/protocols/433.92/tfa2017.c [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+	// options_add(&tfa2017->options, "0", "decimals", OPTION_HAS_VALUE, DEVICES_SETTING, JSON_NUMBER, (void *)1, "[0-9]");
+	options_add(&tfa2017->options, "0", "temperature-decimals", OPTION_HAS_VALUE, GUI_SETTING, JSON_NUMBER, (void *)1, "[0-9]");
+	options_add(&tfa2017->options, "0", "humidity-decimals", OPTION_HAS_VALUE, GUI_SETTING, JSON_NUMBER, (void *)1, "[0-9]");
+	options_add(&tfa2017->options, "0", "humidity-offset", OPTION_HAS_VALUE, DEVICES_SETTING, JSON_NUMBER, (void *)0, "[0-9]");
+	options_add(&tfa2017->options, "0", "temperature-offset", OPTION_HAS_VALUE, DEVICES_SETTING, JSON_NUMBER, (void *)0, "[0-9]");
+	options_add(&tfa2017->options, "0", "show-humidity", OPTION_HAS_VALUE, GUI_SETTING, JSON_NUMBER, (void *)1, "^[10]{1}$");
+	options_add(&tfa2017->options, "0", "show-temperature", OPTION_HAS_VALUE, GUI_SETTING, JSON_NUMBER, (void *)1, "^[10]{1}$");
+
+
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/tfa2017.c]: Can you remove this bogus new line?

File: libs/pilight/protocols/433.92/tfa2017.c [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	}
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void tfa2017Init(void) {
+	protocol_register(&tfa2017);
+	protocol_set_id(tfa2017, "tfa2017");
+	protocol_device_add(tfa2017, "tfa2017", "TFA 30.X Temp Hum Sensor Protocol 2017");
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/tfa2017.c]: Is the addition of protocol `Protocol 2017` necessary?

File: tests/protocols/tfa2017.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	{
+		"4752 1032 912 1039 436 613 841 594 373 1056 431 532 443 535 436 537 927 552 415 558 418 561 416 1037 920 1033 922 555 420 557 416 1034 923 557 424 546 436 1014 453 516 944 535 435 546 444 998 473 506 944 1012 943 1007 475 496 962 526 446 526 446 996 486 490 488 489 962 517 465 516 457 514 466 510 465 985 495 497 490 472 498 476 504 488 491 473 495 475 507 467 509 473 505 465 507 496 961 973 976 966 987 496 488 486 496 946 1012 940 536 438 1027 453 526 921 547 436 540 426 568 408 1051 435 534 445 517 455 530 921 1027 931 1014 465 526 447 528 921 1034 446 526 455 516 931 547 428 1025 461 515 465 514 934 558 414 1045 908 1015 940 540 435 1023 458 509 480 505 931 556 414 567 416 1023 455 528 448 525 456 517 456 517 932 547 436 535 435 557 417 558 425 546 426 546 435 536 441 534 445 539 434 537 445 1004 942 1035 917 1030 459 511 457 517 933 1015 941 537 445 1004 468 517 933 547 435 554 419 543 434 1023 469 500 470 499 476 504 942 1025 924 1030 451 516 466 505 942 1013 466 506 476 496 952 536 435 1017 468 510 468 506 941 528 447 1015 940 1003 959 521 446 1013 465 518 456 517 933 537 446 530 448 1001 477 499 475 509 472 496 476 507 506 297579",
```
Comment (CurlyMoo) [inline tests/protocols/tfa2017.h]: Maybe add a test with a negative temperature?

File: libs/pilight/protocols/433.92/tfa2017.c (hunk 2 above)
Comment (behrisch) [reply libs/pilight/protocols/433.92/tfa2017.c]: I just wanted to give it a description different from the string which is in tfa30. Suggestions are welcome.

Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/tfa2017.c]: Maybe `tfa3.0-2`?
And `TFA 30.X Temp Hum Sensor Protocol v2`?

Comment (behrisch) [reply libs/pilight/protocols/433.92/tfa2017.c]: I am not sure, I think 3.0 is misleading because as far as I understood the 30 refers to the first two digits of the product number (Kat.-Nr.?) as printed on the device which is 30.3208.02 for my sender and 30.3054.10 on the receiver. I will ask in the forum threads maybe there is a good idea for a name and description.

File: tests/protocols/tfa2017.h (hunk 3 above)
Comment (behrisch) [reply tests/protocols/tfa2017.h]: Will do as soon as it is cold enough, because all my tests are based on real measurements.

File: libs/pilight/protocols/433.92/tfa2017.c (hunk 1 above)
Comment (behrisch) [reply libs/pilight/protocols/433.92/tfa2017.c]: OK, can I add commits to this PR or do you want a new one which is squashed?

Comment (CurlyMoo) [reply libs/pilight/protocols/433.92/tfa2017.c]: Just push to the current branch. Sqaushing can be done in git when merging.

**Comment by behrisch** [comment]
I added more tests and removed the one line. No answers on the forum about the name though (and there is again one unit tests failing which seems to be non-deterministic).

File: libs/pilight/protocols/433.92/tfa2017.c [hunk 4]
```diff
@@ -130,9 +130,12 @@ static void parseCode(char **message) {
 		}
 		tmp = tmp->next;
 	}
+	if (humidity<0. || humidity>100.) {
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/tfa2017.c]: Can you make this:
```
if(humidity < 0 || humidity > 100) {
```

File: libs/pilight/protocols/433.92/tfa2017.c [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		if (long_pulse==2 && (prev==20 || (x>7 && prev==x-1))) {
+			if (s==0 || i > start[s-1]+MESSAGE_LENGTH) {
+				start[s++] = i;
+				prev = 0;
+			}
+		}
+	}
+	// The protocol sends the message three times in a row.
+	// If we find two identical ones, we consider it valid.
+	if (s<2) {
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/tfa2017.c]: Please check the pilight style guide.

File: tests/protocols/tfa2017.h [hunk 6]
```diff
@@ -1,14 +1,32 @@
...
+		"{\"id\":1,\"temperature\":2.50,\"humidity\":93.00}",
+		NULL,
+		0
+	},
+	{ // error raw length too short
```
Comment (CurlyMoo) [inline tests/protocols/tfa2017.h]: Nice

**Comment by CurlyMoo** [comment]
All checks succeeded. Good work.

**Comment by behrisch** [comment]
Thanks. I hope it will be useful.

**Comment by CurlyMoo** [comment]
Can you change the license to MPLv2?

File: libs/pilight/protocols/433.92/tfa2017.c [hunk 7]
```diff
@@ -1,19 +1,9 @@
 /*
 	Copyright (C) 2014, 2018 CurlyMo & DonBernos & Michael Behrisch
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/tfa2017.c]: And of course make sure the indenting matches :smile:

**Comment by CurlyMoo** [comment]
Can you fix this warning?
https://travis-ci.org/pilight/pilight/builds/478696545#L901
```
 snprintf(*message, 255, "");
```

change it to:
```
memset(*message, 0, 255);
```
or maybe:
```
memset(&(*message), 0, 255);
```

**Comment by CurlyMoo** [comment]
Ready for merging?

**Comment by behrisch** [comment]
I am, if you are :-)

**Comment by CurlyMoo** [comment]
Thanks for this wonderful merge request!

**Comment by behrisch** [comment]
Thank you for integrating! I had a short look at the checksum code in https://github.com/adilosa/weathermon/blob/master/weathermon.c which may be adaptable to this protocol but my first try did not work out.
